### DISCUSSION TITLE
I have a question, how to render the video stream in the browser?

### DISCUSSION THREAD

**Discussion by zw-999** [yume-chan/ya-webadb#699] [General]
Can @yume-chan/scrcpy-decoder-tinyh264 and @yume-chan/scrcpy-decoder-webcodecs be used separately?  Since I'm not using nodejs, I'm having a problem rendering the video stream in the browser.

**Comment by yume-chan** [comment]
`@yume-chan/scrcpy-decoder-tinyh264` and `@yume-chan/scrcpy-decoder-webcodecs` can only be used in browsers. 

They only used several types in `@yume-chan/scrcpy` to define the input format. If you can supply your video data in a compatible format, you can use them to decode and render your own videos.

**Reply by zw-999** [reply]
`const client = await AdbScrcpyClient.start(
        adb,
        DEFAULT_SERVER_PATH,
        VERSION,
        new AdbScrcpyOptionsLatest(
          new ScrcpyOptionsLatest({
            scid: ScrcpyInstanceId.random(),
            audio: false,
            lockVideoOrientation: ScrcpyVideoOrientation1_18.Portrait,
            videoBitRate: 2_000_000,
            videoCodecOptions: new CodecOptions({
              profile: AndroidAvcProfile.Baseline,
              level: AndroidAvcLevel.Level4,
            }),
            cleanup: false,
            tunnelForward: true,
          })
        )
      );`
This is the code I saw in your demo. Can the stream I read from the video after using this command be rendered using @yume-chan/scrcpy-decoder-tinyh264?
`adb shell CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 2.4 scid=28971338 video_bit_rate=3000000 video_codec_options=profile=1,level=2048`

**Reply by yume-chan** [reply]
You need to connect to Scrcpy server (https://github.com/Genymobile/scrcpy/blob/master/doc/develop.md#connection and https://docs.tangoapp.dev/scrcpy/connect-server/), and send the video stream to your Web app (for example with `fetch` or WebSocket). Exactly how to do those depends on the language and framework you use.

Then you need to convert the video stream into a `ReadableStream` (with `fetch` it's already a `ReadableStream`, with WebSocket you can use https://docs.tangoapp.dev/tango/web-stream/websocket/), parse the stream into video frames (https://docs.tangoapp.dev/scrcpy/video/#with-yume-chanscrcpy), then pipe those frames into the decoder (https://docs.tangoapp.dev/scrcpy/video/tiny-h264/#usage)

**Reply by zw-999** [reply]
I set up a local server to forward the video stream of scrcpy. I got the video stream through fetch and will try to render it, but my browser interface does not display correctly.  Here's my code and the scrcpy log.how do I modify my code?
```
scrcpy 2.4

  "dependencies": {
    "@yume-chan/scrcpy": "^0.0.24",
    "@yume-chan/scrcpy-decoder-tinyh264": "^0.0.24",
}
```
```
import { TinyH264Decoder } from "@yume-chan/scrcpy-decoder-tinyh264";
let rsp =  await fetch("http://127.0.0.1:5000/video_stream")
let video = rsp.body
import { ScrcpyOptionsLatest } from "@yume-chan/scrcpy";
const options = new ScrcpyOptionsLatest({});
const { metadata, stream } = await options.parseVideoStreamMetadata(video);
// const videoPacketStream = stream.pipeThrough(options.createMediaStreamTransformer());
console.log(metadata)
console.log(stream)
const decoder = new TinyH264Decoder();
document.body.appendChild(decoder.renderer);
stream.pipeTo(decoder.writable).catch((err) => {console.log(err)});
```
```
➜  scrcpy-2.4 cat start.sh
# /usr/bin/bash
echo "start"
adb push x/server/scrcpy-server /data/local/tmp/scrcpy-server.jar
adb reverse localabstract:scrcpy_28971338 tcp:27183
adb shell CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 2.4 scid=28971338 video_bit_rate=3000000 max_fps=30 video_codec_options=profile=1,level=2048

output:
x/server/scrcpy-server: 1 file pushed, 0 skipped. 70.0 MB/s (69007 bytes in 0.001s)
[server] INFO: Device: [HUAWEI] HUAWEI ELE-AL00 (Android 10)
[server] WARN: Audio disabled: it is not supported before Android 11
[server] DEBUG: Audio encoder stopped
[server] DEBUG: Using video encoder: 'OMX.hisi.video.encoder.avc'
[server] DEBUG: Video codec option set: profile (Integer) = 1
[server] DEBUG: Video codec option set: level (Integer) = 2048
[server] DEBUG: Display: using SurfaceControl API
```
This is what the browser shows.
![p1](https://github.com/user-attachments/assets/1999d0cc-66b5-4903-a51f-5b1b22433d2b)

**Reply by yume-chan** [reply]
> ```
> // const videoPacketStream = stream.pipeThrough(options.createMediaStreamTransformer());
> ```

This line is needed. You should pipe `videoPacketStream` into `decoder.writable`. Otherwise the code looks correct (and `parseVideoStreamMetadata` is working). You can also read `videoPacketStream` yourself instead of piping it into the decoder to check if the packets are flowing.

**Reply by zw-999** [reply]
The console reported an error after adding this line。
````
import { TinyH264Decoder } from "@yume-chan/scrcpy-decoder-tinyh264";
let rsp =  await fetch("http://127.0.0.1:5000/video_stream")
let video = rsp.body
import { ScrcpyOptionsLatest } from "@yume-chan/scrcpy";
const options = new ScrcpyOptionsLatest({});
const { metadata, stream } = await options.parseVideoStreamMetadata(video);
const videoPacketStream = stream.pipeThrough(options.createMediaStreamTransformer());
console.log(metadata)
console.log(stream)
const decoder = new TinyH264Decoder();

document.body.appendChild(decoder.renderer);
videoPacketStream.pipeTo(decoder.writable).catch((err) => {console.log(err)});
````
````
{deviceName: 'ELE-AL00', codec: 1748121140, width: 1080, height: 2336}    main.js:17
PushReadableStream {locked: true}    main.js:18 
TypeError: YuvBuffer.format is not a function   main.js:25
    at TinyH264Decoder.configure_fn (@yume-chan_scrcpy-decoder-tinyh264.js?v=445583b4:241:28)
    at async Object.write (@yume-chan_scrcpy-decoder-tinyh264.js?v=445583b4:184:13)
````

**Reply by yume-chan** [reply]
Vite can't load version 0.0.24. Try version 1.0.0 I released today.

**Reply by zw-999** [reply]
With no error after a 1.0.0 release, but no video rendering out, according to the document I tried many ways to "https://docs.tangoapp.dev/scrcpy/video/tiny-h264/" but in the end also didn't apply colours to a drawing, can you have a similar example reference?  Thank you very much. This has been bothering me for a long time. 
My local environment:
````
cpu  M1
node    v22.12.0
scrcpy 2.4

"@yume-chan/scrcpy": "^1.0.0",
"@yume-chan/scrcpy-decoder-tinyh264": "^1.0.0",
"@yume-chan/scrcpy-decoder-webcodecs": "^1.0.0",
````
I start the command scrcpy-server `adb shell CLASSPATH=/data/local/tmp/scrcpy-server.jar app_process / com.genymobile.scrcpy.Server 2.4 scid=28971338 video_bit_rate=2000000 max_fps=30 video_codec_options=profile=1,level=2048`
and my code
````
import { TinyH264Decoder } from "@yume-chan/scrcpy-decoder-tinyh264";
import { ScrcpyOptions2_4 } from "@yume-chan/scrcpy";

let rsp =  await fetch("http://127.0.0.1:5000/video_stream")
let video = rsp.body
const options = new ScrcpyOptions2_4({});
const { metadata, stream } = await options.parseVideoStreamMetadata(video);
const videoPacketStream = stream.pipeThrough(options.createMediaStreamTransformer());
console.log(metadata)
console.log(stream)
const decoder = new TinyH264Decoder();
videoPacketStream.pipeTo(decoder.writable).catch((e) => {console.error(e)});
document.body.appendChild(decoder.renderer);
console.log("ok")
````
<img width="1407" alt="image" src="https://github.com/user-attachments/assets/8e7a92dc-6786-42a8-abf1-367a7bce499e">

**Reply by yume-chan** [reply]
For example, this project uses WebSocket to transfer multiple streams and render them using TinyH264 decoder: https://github.com/project-SIMPLE/simple.webplatform/blob/36cbef318e17300e5ea5400cb48a4f88a54ba8e1/src/components/WebSocketManager/VideoStreamManager.tsx#L80

When 504 happens Vite should automatically reload the page, if it didn't, or if it still returns 504 even after reloading manually, then this is an issue in Vite.

**Reply by yume-chan** [reply]
Update: It's a Vite bug https://github.com/vitejs/vite/issues/8427
